File: tools/bench/bounded_queue_busywait.cpp
// 바쁜 대기와 조건변수의 CPU 비용 (C++).
//
// Python 판(bounded_queue_busywait.py)과 같은 실험이다. 다른 것은 GIL 이 없다는 점
// 하나이고, 그 하나가 결과의 성격을 바꾼다 — C++ 의 회전 소비자는 남의 코어를
// 뺏지 않으므로 벽시계는 그대로이고 CPU 만 탄다. Python 은 벽시계까지 나빠진다.
//
// 측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / g++ 13.3.0 -std=c++17 -O2 / 4 코어).
// 빌드·실행: g++ -std=c++17 -O2 -pthread tools/bench/bounded_queue_busywait.cpp -o /tmp/bw && /tmp/bw

#include <sys/resource.h>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <deque>
#include <mutex>
#include <thread>
#include <vector>
using namespace std;
using namespace std::chrono;

static const int ITEMS = 200;
static const int GAP_MS = 2;
static const int REPEAT = 5;

double cpu_seconds() {
    rusage ru{};
    getrusage(RUSAGE_SELF, &ru);
    return ru.ru_utime.tv_sec + ru.ru_utime.tv_usec / 1e6 +
           ru.ru_stime.tv_sec + ru.ru_stime.tv_usec / 1e6;
}

struct Result { double wall, cpu; };

Result run_busy() {
    deque<int> box;
    mutex mu;
    atomic<bool> done{false};
    int got = 0;

    auto t0 = steady_clock::now();
    double c0 = cpu_seconds();
    thread p([&] {
        for (int i = 0; i < ITEMS; i++) {
            this_thread::sleep_for(milliseconds(GAP_MS));
            lock_guard<mutex> g(mu);
            box.push_back(i);
        }
        done = true;
    });
    thread c([&] {
        for (;;) {
            {
                lock_guard<mutex> g(mu);
                if (!box.empty()) { box.pop_front(); got++; continue; }
            }
            if (done && box.empty()) return;
            // 아무 일도 안 하지만 루프는 전속력으로 돈다
        }
    });
    p.join(); c.join();
    return {duration<double>(steady_clock::now() - t0).count(), cpu_seconds() - c0};
}

Result run_cv() {
    deque<int> box;
    mutex mu;
    condition_variable cv;
    bool done = false;
    int got = 0;

    auto t0 = steady_clock::now();
    double c0 = cpu_seconds();
    thread p([&] {
        for (int i = 0; i < ITEMS; i++) {
            this_thread::sleep_for(milliseconds(GAP_MS));
            { lock_guard<mutex> g(mu); box.push_back(i); }
            cv.notify_one();
        }
        { lock_guard<mutex> g(mu); done = true; }
        cv.notify_all();
    });
    thread c([&] {
        for (;;) {
            unique_lock<mutex> lk(mu);
            cv.wait(lk, [&] { return !box.empty() || done; });   // 락을 놓고 잠든다
            if (!box.empty()) { box.pop_front(); got++; continue; }
            return;
        }
    });
    p.join(); c.join();
    return {duration<double>(steady_clock::now() - t0).count(), cpu_seconds() - c0};
}

void survey(Result (*fn)(), const char* name) {
    vector<double> w, c;
    for (int i = 0; i < REPEAT; i++) { Result r = fn(); w.push_back(r.wall); c.push_back(r.cpu); }
    sort(w.begin(), w.end()); sort(c.begin(), c.end());
    printf("%-12s 벽시계 중앙값 %.3f초 (%.3f~%.3f)   CPU 중앙값 %.3f초 (%.3f~%.3f)   점유율 %.0f%%\n",
           name, w[REPEAT / 2], w.front(), w.back(),
           c[REPEAT / 2], c.front(), c.back(), c[REPEAT / 2] / w[REPEAT / 2] * 100);
}

int main() {
    printf("항목 %d개, 생산 간격 %dms, %d회 실행, 코어 4\n", ITEMS, GAP_MS, REPEAT);
    survey(run_busy, "바쁜 대기");
    survey(run_cv, "조건변수");
    return 0;
}

File: tools/bench/bounded_queue_busywait.py
"""바쁜 대기와 조건변수의 CPU 비용 (Python).

유계 큐가 비었을 때 소비자는 기다려야 한다. 기다리는 방법이 둘이다.
  1. 바쁜 대기 — `while queue.empty(): pass`. 큐를 계속 들여다본다.
  2. 조건변수 — `cv.wait()`. 커널이 스레드를 재우고 통지가 올 때 깨운다.

둘 다 "기다린다" 지만 벽시계 시간이 아니라 **CPU 시간**을 재면 갈린다.
바쁜 대기는 아무 일도 안 하면서 코어를 하나 통째로 태운다.

측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / CPython 3.13.12 / 4 코어).
실행: python3.13 tools/bench/bounded_queue_busywait.py
"""

import statistics
import threading
import time

ITEMS = 200
GAP = 0.002          # 생산자가 한 개 만드는 데 걸리는 시간(느린 상류를 흉내낸다)
REPEAT = 5


def cpu_time():
    """이 프로세스가 실제로 CPU 를 쥔 시간. 벽시계와 다르다."""
    t = time.process_time()
    return t


def run_busy():
    box = []
    done = [False]
    got = [0]

    def producer():
        for i in range(ITEMS):
            time.sleep(GAP)
            box.append(i)
        done[0] = True

    def consumer():
        while True:
            if box:
                box.pop(0)
                got[0] += 1
            elif done[0]:
                return
            # 아무 일도 안 하지만 루프는 전속력으로 돈다

    p = threading.Thread(target=producer)
    c = threading.Thread(target=consumer)
    t0, c0 = time.perf_counter(), cpu_time()
    p.start(); c.start(); p.join(); c.join()
    return time.perf_counter() - t0, cpu_time() - c0, got[0]


def run_cv():
    box = []
    done = [False]
    got = [0]
    cv = threading.Condition()

    def producer():
        for i in range(ITEMS):
            time.sleep(GAP)
            with cv:
                box.append(i)
                cv.notify()
        with cv:
            done[0] = True
            cv.notify_all()

    def consumer():
        while True:
            with cv:
                while not box and not done[0]:
                    cv.wait()          # 락을 놓고 잠든다
                if box:
                    box.pop(0)
                    got[0] += 1
                    continue
                return

    p = threading.Thread(target=producer)
    c = threading.Thread(target=consumer)
    t0, c0 = time.perf_counter(), cpu_time()
    p.start(); c.start(); p.join(); c.join()
    return time.perf_counter() - t0, cpu_time() - c0, got[0]


def survey(fn, name):
    walls, cpus = [], []
    for _ in range(REPEAT):
        w, c, got = fn()
        assert got == ITEMS, got
        walls.append(w); cpus.append(c)
    print(f"{name:10s} 벽시계 중앙값 {statistics.median(walls):.3f}초 "
          f"({min(walls):.3f}~{max(walls):.3f})   "
          f"CPU 중앙값 {statistics.median(cpus):.3f}초 "
          f"({min(cpus):.3f}~{max(cpus):.3f})   "
          f"점유율 {statistics.median(cpus) / statistics.median(walls) * 100:.0f}%")


if __name__ == "__main__":
    print(f"항목 {ITEMS}개, 생산 간격 {GAP * 1000:.0f}ms, {REPEAT}회 실행, 코어 4")
    survey(run_busy, "바쁜 대기")
    survey(run_cv, "조건변수")
